controllers/terraform_controller.go

```diff
@@ -487,33 +496,82 @@ func (r *TerraformReconciler) reconcile(ctx context.Context, runnerClient runner
 		log.Info("should apply == false")
 	}
 
-	if r.shouldWriteOutputs(terraform, outputs) {
-		terraform, err = r.writeOutput(ctx, terraform, runnerClient, outputs, revision)
+	terraform, err = r.processOutputs(ctx, runnerClient, terraform, tfInstance, revision)
+	if err != nil {
+		return terraform, err
+	}
+
+	if r.shouldDoHealthChecks(terraform) {
+		terraform, err = r.doHealthChecks(ctx, terraform, runnerClient)
 		if err != nil {
 			return terraform, err
 		}
 
 		if err := r.patchStatus(ctx, objectKey, terraform.Status); err != nil {
-			log.Error(err, "unable to update status after writing outputs")
+			log.Error(err, "unable to update status after doing health checks")
 			return terraform, err
 		}
 	}
 
-	if r.shouldDoHealthChecks(terraform) {
-		terraform, err = r.doHealthChecks(ctx, terraform, runnerClient)
+	return terraform, nil
+}
+
+func (r *TerraformReconciler) processOutputs(ctx context.Context, runnerClient runner.RunnerClient, terraform infrav1.Terraform, tfInstance string, revision string) (infrav1.Terraform, error) {
+
+	log := ctrl.LoggerFrom(ctx)
+	objectKey := types.NamespacedName{Namespace: terraform.Namespace, Name: terraform.Name}
+
+	outputs := map[string]tfexec.OutputMeta{}
+	var err error
+	terraform, err = r.obtainOutputs(ctx, terraform, tfInstance, runnerClient, revision, &outputs)
+	if err != nil {
+		return terraform, err
+	}
```

Comment:
It would make more sense to create the outputs map inside `obtainOutputs` and return it from the from. Likewise, I would also pass the `terraform` struct as a pointer rather than passing and returning it.

Reply:
nop not me.
chaos tests has been added by @tomhuang12 a while ago. I just make it less CPU-consuming, but it's still meaningful enough for e2e.

Reply:
I did think that as well.

But this pattern is generally suggested by k8sclient, controller SDK, as well as KS controller. so I think we can just leave it there to conform to the overall writing style.